### PR TITLE
Unit tests for native histograms

### PR DESCRIPTION
**Pull request by zeitlinger**
Relates #11170, #12146

Supercedes https://github.com/prometheus/prometheus/pull/12080

### CODE REVIEW

File: promql/parser/generated_parser.y
Comment (beorn7) [inline promql/parser/generated_parser.y:669]: Nit: Throughout the Prometheus code base, we prefer correct English syntax in comments. (I.e. start sentence with a capital letter, end with a period. Apply everywhere.)

File: promql/parser/generated_parser.y
Comment (beorn7) [inline promql/parser/generated_parser.y:684]: I'm wondering if the naming should be more consistent with `series_value` above. Something like `histogram_series_value`

**Review by beorn7 — COMMENTED** [review]
I just took a quick look to find out what we discussed in the sync meeting. Indeed, the grammar for the test series was already in the generated parser. So all is good.

I couldn't resist and left a few nitpicks. Will do a full review later when I have cleaned up my backlog of even older reviews.

File: promql/parser/lex.go
Comment (beorn7) [inline promql/parser/lex.go:32]: Despite this if statement, calling this function isn't free. In particular, the arguments are stringified before even reaching this. I'm not sure how performance critical the lexing is (some users have veeerrry long rules files), but conservatively, you should probably remove these calls before merging.

File: promql/parser/parse.go
Comment (beorn7) [inline promql/parser/parse.go:470]: Go doc comment style (which I think should be followed even if the function is not exporter): "newMap is used when building the FloatHistogram from a map."

File: promql/parser/parse.go
Comment (beorn7) [inline promql/parser/parse.go:475]: As above:
This function is used... → mergeMaps is used…

File: promql/parser/parse.go
Comment (beorn7) [inline promql/parser/parse.go:504]: End with period.

File: promql/parser/parse.go
Comment (beorn7) [inline promql/parser/parse.go:520]: This is used ... → buildHistogramFromMap is used ...

File: promql/parser/parse_test.go
Comment (beorn7) [inline promql/parser/parse_test.go:24]: Has to go to the bottom most block.

File: promql/testdata/native_histograms.test
Comment (beorn7) [inline promql/testdata/native_histograms.test:15]: Maybe do an `eval instant at 5m empty_histogram` to prove that the histogram is even there.

File: promql/testdata/native_histograms.test
Comment (beorn7) [inline promql/testdata/native_histograms.test:5]: The empty `{}` for label selections aren't required. I would usually remove them.

File: promql/testdata/native_histograms.test
Comment (beorn7) [inline promql/testdata/native_histograms.test:21]: Maybe use a different value for `sum` to make sure we are seeing the right values below.

File: promql/testdata/native_histograms.test
Comment (beorn7) [inline promql/testdata/native_histograms.test:116]: We should probably allow the opposite direction as well in the test framework (even if we have to invert things internally).

File: promql/testdata/native_histograms.test
Comment (beorn7) [inline promql/testdata/native_histograms.test:108]: I would also test the full histogram (i.e. `eval instant at 50m incr_histogram`).

File: promql/testdata/native_histograms.test
Comment (beorn7) [inline promql/testdata/native_histograms.test:128]: Same as before, let's test the full histogram.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Great stuff. Mostly style comments here.

We can have way more test cases, for all the arithmetic and rate and such. But do those at your leisure. To have the framework in place is most important so that we can add more tests at will. The existing tests are already very nice.

File: model/histogram/float_histogram.go
Comment (zeitlinger) [reply model/histogram/float_histogram.go:164]: Happy to fix it - do you have an idea?

File: promql/testdata/native_histograms.test
Comment (skl) [reply promql/testdata/native_histograms.test:19]: Looks like the parser doesn't support this yet. I added some todo's with the failing tests in 3eb0735a7 against all suggestions of full histogram queries.

File: promql/testdata/native_histograms.test
Comment (skl) [reply promql/testdata/native_histograms.test]: Removed all empty label selectors in 3eb0735a7.

File: promql/testdata/native_histograms.test
Comment (skl) [reply promql/testdata/native_histograms.test]: Changed sum value and subsequent assertion in 3eb0735a7.

File: promql/testdata/native_histograms.test
Comment (skl) [reply promql/testdata/native_histograms.test:116]: Added in 3eb0735a7, see above regarding parser support.

File: promql/testdata/native_histograms.test
Comment (skl) [reply promql/testdata/native_histograms.test:140]: Added in 3eb0735a7, see above regarding parser support.

File: model/histogram/float_histogram.go
Comment (skl) [inline model/histogram/float_histogram.go:138]: Looks like this causes the following for previous Go version tests:

```
model/histogram/float_histogram.go:138:23: undefined: strings.CutSuffix
```

Comment (beorn7) [reply model/histogram/float_histogram.go:138]: Luckily, we don't need the trim function, see my comment above.

File: model/histogram/float_histogram.go
Comment (beorn7) [reply model/histogram/float_histogram.go:164]: You could fill the gaps between spans with empty buckets.

File: model/histogram/float_histogram.go
Comment (zeitlinger) [reply model/histogram/float_histogram.go]: Funnily, copilot suggested `%g` - I just ignored it...

File: model/histogram/float_histogram.go
Comment (zeitlinger) [reply model/histogram/float_histogram.go:170]: good idea - done.

File: promql/testdata/native_histograms.test
Comment (skl) [reply promql/testdata/native_histograms.test:18]: Added in 65cbd1932

File: promql/testdata/native_histograms.test
Comment (skl) [reply promql/testdata/native_histograms.test:137]: Fixed in 65cbd1932

File: promql/testdata/native_histograms.test
Comment (skl) [inline promql/testdata/native_histograms.test:217]: @beorn7 please can you check my understanding of this test case? I thought median would be zero given an equal distribution about zero. A positive response of `0.5` to me implies the distribution is skewed above zero?

File: model/histogram/float_histogram_test.go
Comment (beorn7) [inline model/histogram/float_histogram_test.go:985]: That's not correct. In the original histogram, the second span has an offset to the previous span of 3. That's reduced here to only 1. Or in other word: The test was correct before.

File: model/histogram/float_histogram_test.go
Comment (beorn7) [inline model/histogram/float_histogram_test.go:998]: That's not correct, either. Let's talk…

**Review by beorn7 — CHANGES_REQUESTED** [review]

File: promql/testdata/native_histograms.test
Comment (beorn7) [reply promql/testdata/native_histograms.test:225]: See discussion on Slack. If you add a single sample in the zero bucket, it should come out as zero.
The result `0.5` is technically correct.

**Comment by beorn7** [comment]
I believe the “fix” of the histogram compaction is wrong. See also discussion on Slack. Let's first clarify that issue. If I'm right, then I would like to ask you to remove the corresponding changes in the tests and the code, after which it will be easier for me to do the final review.

File: promql/testdata/native_histograms.test
Comment (beorn7) [reply promql/testdata/native_histograms.test:225]: I'm also pretty sure that classic histograms have the same behavior.

Overall, I think the issue is fairly academic because "real world" histograms will have many observations, and they are certainly not distributed in a symmetric fashion where you would mind if the median is really the midpoint between the two observations around the median. Plus, it only matters if there are empty buckets in that range, which is fairly unlikely in practice.

Comment (beorn7) [reply promql/testdata/native_histograms.test:225]: Or in other words: Let's not "fix" this behavior right now. We can think about it separately. Having the test here is good to demonstrate the behavior (and it deserves a comment along the lines: "If the quantile happens to be located in a span of empty buckets, the actually returned value is the lower bound of the first populated bucket after the span of empty buckets."

Comment (skl) [reply promql/testdata/native_histograms.test]: Set expected result to 0.5 in 95446ec73

**Comment by skl** [comment]
Restored history compaction in 10988f35c, agree with @beorn7 on incorrect offsets.

cc @zeitlinger - I restored compaction implementation from `main` and did my best to merge with your work.

**Comment by zeitlinger** [comment]
> Restored history compaction in [10988f3](https://github.com/prometheus/prometheus/commit/10988f35c7e9d130f5692667c75d11b28df7522d), agree with @beorn7 on incorrect offsets.
> 
> cc @zeitlinger - I restored compaction implementation from `main` and did my best to merge with your work.

looks good - @beorn7 do you want to give it a final reiew?

File: cmd/promtool/unittest.go
Comment (beorn7) [inline cmd/promtool/unittest.go:372]: Move this whole section after the next block (`if err != nil { ... }`). In that way, you don't have to gate on the error here.

**Comment by zeitlinger** [comment]
was this comment deleted?

File: model/histogram/float_histogram.go
Comment (beorn7) [inline model/histogram/float_histogram.go:844]: Do we really need this change? The invariant that targetSchema must be ≤ the schema of FloatHistogram has to be enforced by the code. If this error ever happens, it is a programming error. Therefore, it is fine to panic, and we don't have to propagate the error everywhere. (As you can see in this PR, it makes the code much mode fiddly.)

File: model/histogram/float_histogram.go
Comment (beorn7) [inline model/histogram/float_histogram.go:250]: Please don't add an error return here, see comment below.

**Review by beorn7 — CHANGES_REQUESTED** [review]
See my comment about the additional error return. I don't think we need it, and it creates a lot of noise in this PR. Either convince me that this error return is really needed, or remove it again. (I hope for the latter and will finish the review afterwards because it will make it much simpler.)

**Comment by beorn7** [comment]
> was this comment deleted?

Yes, it was meant to be sent as part of the review, not as an individual comment.

File: cmd/promtool/unittest.go
Comment (zeitlinger) [reply cmd/promtool/unittest.go:372]: this doesn't work, because this block can also set `err`, e.g. `err = parseErr`, which is then processed by the next block

File: model/histogram/float_histogram.go
Comment (zeitlinger) [reply model/histogram/float_histogram.go:148]: yes, and the compact statement below merges all the buckets (positive and negative) anyways.

**Comment by zeitlinger** [comment]
> I don't think it's done. Also, the panic is still there.

(couldn't reply to the original comment)

```
		if len(spans) > 1 {
			panic(fmt.Sprintf("histogram with multiple %s spans not supported", kind))
		}
```

from https://github.com/zeitlinger/prometheus/blob/b5ffba16d74f8c9a7d7c4343a56babbb28bda26f/model/histogram/float_histogram.go#L170

This panic can only happen if `m.Compact(math.MaxInt)` produces more than a single bucket.

File: model/histogram/float_histogram.go
Comment (zeitlinger) [reply model/histogram/float_histogram.go:844]: yes, this is a difficult one. 

When we parse a histogram series expression, this is a user error instead of a programmer error, but I also agree that the code is getting more complex as a result.

This is the call site:

```go
	return p.histogramsSeries(base, inc, times, func(a, b *histogram.FloatHistogram) (*histogram.FloatHistogram, error) {
		return a.Add(b)
	})
```

Before the signature change, I had

```go
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("recovered:", r)
			err = fmt.Errorf("error combining histograms: %v", r)
		}
	}()
```

Alternatives I can think of:
1. Use the recover again
2. Check for schema explicitly before calling add in histogramsSeries (possibly adding a `canBeCombined` method)
3. `TryAdd` and `Add` (first one returns an error)

**Comment by zeitlinger** [comment]
@beorn7 addressed easy comments and responded to the others to discuss options

File: cmd/promtool/unittest.go
Comment (beorn7) [reply cmd/promtool/unittest.go:372]: I see. Then I guess it's fine as is.

File: model/histogram/float_histogram.go
Comment (beorn7) [inline model/histogram/float_histogram.go:141]: Could you add a comment here that this is to reduce the number of spans to one?

File: model/histogram/float_histogram.go
Comment (beorn7) [reply model/histogram/float_histogram.go:161]: With your explanation (given elsewhere) I understand now how the problem got solved. See above, please add a comment to `m.Compact(math.MaxInt)` so that the reader understands this little trick.

File: model/histogram/float_histogram.go
Comment (beorn7) [reply model/histogram/float_histogram.go:835]: I see two options. I'll comment on the call site below.

File: promql/parser/parse.go
Comment (beorn7) [inline promql/parser/parse.go:490]: To avoid the error (or panic), you can do the following to make the addition always work:
```suggestion
	return p.histogramsSeries(base, inc, times, func(a, b *histogram.FloatHistogram) (*histogram.FloatHistogram, error) {
		if b.Schema >= a.Schema {
			return a.Add(b)
		}
		return b.Add(a)
```

Or if you _want_ an error, do the following:

```suggestion
	return p.histogramsSeries(base, inc, times, func(a, b *histogram.FloatHistogram) (*histogram.FloatHistogram, error) {
		if b.Schema < a.Schema {
			return nil, fmt.Errorf("some meaningful error message")
		}
		return a.Add(b)
```

**Review by beorn7 — CHANGES_REQUESTED** [review]

File: model/histogram/float_histogram.go
Comment (zeitlinger) [reply model/histogram/float_histogram.go]: OK, did as advised.

**Comment by beorn7** [comment]
OK, I'm back to full review mode then. :)

File: promql/testdata/native_histograms.test
Comment (beorn7) [inline promql/testdata/native_histograms.test:50]: Would be good to have sum and count here different, too.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Excellent. Just two minor nits.

And then the question how to squash the 57 commits. If you feel like it, you could squash them into a smaller number of commits that make sense (e.g. creating of the test framework, adding tests, etc.). But that might be a lot of work. I'm also fine if I just squash everything into one commit, co-authored by all three of you.

**Comment by zeitlinger** [comment]
> I'm also fine if I just squash everything into one commit, co-authored by all three of you.

Yes, that would be great - can you do that when hitting the merge button?

**Review by beorn7 — APPROVED** [review]

**Comment by beorn7** [comment]
Thank you so much @hdost @zeitlinger @skl . This is a huge milestone. Now we can convert a bunch of fiddly code level tests to the testing framework, and it will be very easy to add more tests for existing and future histogram features in PromQL.

File: promql/parser/generated_parser.y.go
Comment (bboreham) [inline promql/parser/generated_parser.y.go:43]: I'd like to note for the record that this PR un-did the formatting applied to this generated file in #9563.

In my opinion this version is correct - we shouldn't be reformatting generated files - but I would have done the formatting change in its own PR for clarity in the history.

Comment (beorn7) [reply promql/parser/generated_parser.y.go:43]: I see. Yeah, generated files either stay unformatted, or the formatting has to be part of the auto-generation.
We should have caught that in #9563…

**Pull request merged**
